src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs

```diff
@@ -104,7 +104,22 @@ protected override ValueTask<RateLimitLease> WaitAsyncCore(int tokenCount, Cance
                 // Don't queue if queue limit reached
                 if (_queueCount + tokenCount > _options.QueueLimit)
                 {
-                    return new ValueTask<RateLimitLease>(CreateFailedTokenLease(tokenCount));
+                    if (_options.QueueProcessingOrder == QueueProcessingOrder.NewestFirst)
+                    {
+                        // remove oldest items from queue until there is space for the newest acquisition request
+                        do
+                        {
+                            RequestRegistration oldestRequest = _queue.DequeueHead();
+                            _queueCount -= oldestRequest.Count;
+                            Debug.Assert(_queueCount >= 0);
+                            oldestRequest.Tcs.TrySetResult(FailedLease);
+                        }
+                        while (_queueCount + tokenCount > _options.QueueLimit);
```

Comment:
Do you think it would be worthwhile to share more logic between the rate limiter implementations?

Reply:
👍 
Yeah it seems like a lot of the logic is common